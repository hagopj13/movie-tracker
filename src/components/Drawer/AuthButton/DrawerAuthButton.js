// @flow
import React from 'react';
import { connect } from 'react-redux';
import { createStructuredSelector } from 'reselect';
import List from '@material-ui/core/List';
import ListItem from '@material-ui/core/ListItem';
import ListItemIcon from '@material-ui/core/ListItemIcon';
import ListItemText from '@material-ui/core/ListItemText';
import LogoutIcon from '@material-ui/icons/Lock';
import LoginIcon from '@material-ui/icons/LockOpen';

import dialogTypes from 'config/dialogTypes';
import authSelectors from 'store/auth/auth.selectors';
import authActions from 'store/auth/auth.actions';
import dialogActions from 'store/ui/dialog/dialog.actions';

type Props = {
  isAuth: boolean,
  onLogout: () => void,
  onShowLoginDialog: () => void,
};

export const DrawerAuthButton = (props: Props) => {
  const { isAuth, onLogout, onShowLoginDialog } = props;

  return (
    <List>
      {isAuth ? (
        <ListItem button onClick={onLogout}>
          <ListItemIcon>
            <LogoutIcon />
          </ListItemIcon>
          <ListItemText primary="Logout" />
        </ListItem>
      ) : (
        <ListItem button onClick={onShowLoginDialog}>
          <ListItemIcon>
            <LoginIcon />
          </ListItemIcon>
          <ListItemText primary="Login" />
        </ListItem>
      )}
    </List>
  );
};

const mapStateToProps = createStructuredSelector({
  isAuth: authSelectors.selectIsAuth,
});

const mapDispatchToProps = {
  onLogout: authActions.logout,
  onShowLoginDialog: () => dialogActions.showDialog(dialogTypes.LOGIN),
};

export default connect(mapStateToProps, mapDispatchToProps)(DrawerAuthButton);
