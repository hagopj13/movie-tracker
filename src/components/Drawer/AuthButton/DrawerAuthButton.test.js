import React from 'react';
import { createShallow } from '@material-ui/core/test-utils';
import ListItem from '@material-ui/core/ListItem';
import ListItemText from '@material-ui/core/ListItemText';

import { DrawerAuthButton } from './DrawerAuthButton';

describe('DrawerAuthButton component', () => {
  let shallow;
  let wrapper;

  const mockShowLoginDialog = jest.fn();
  const mockLogout = jest.fn();

  beforeAll(() => {
    shallow = createShallow();
  });

  beforeEach(() => {
    jest.clearAllMocks();
  });

  describe('isAuth is true', () => {
    beforeEach(() => {
      const mockProps = {
        isAuth: true,
        onLogout: mockLogout,
        onShowLoginDialog: mockShowLoginDialog,
      };
      wrapper = shallow(<DrawerAuthButton {...mockProps} />);
    });

    it('should render the DrawerAuthButton component correctly', () => {
      expect(wrapper).toMatchSnapshot();
    });

    it('should render the logout button', () => {
      expect(wrapper.find(ListItemText).prop('primary')).toBe('Logout');
    });

    it('should call onLogout when the button is clicked', () => {
      wrapper.find(ListItem).simulate('click');
      expect(mockLogout).toHaveBeenCalledTimes(1);
    });
  });

  describe('isAuth is false', () => {
    beforeEach(() => {
      const mockProps = {
        isAuth: false,
        onLogout: mockLogout,
        onShowLoginDialog: mockShowLoginDialog,
      };
      wrapper = shallow(<DrawerAuthButton {...mockProps} />);
    });

    it('should render the DrawerAuthButton component correctly', () => {
      expect(wrapper).toMatchSnapshot();
    });

    it('should render the login button', () => {
      expect(wrapper.find(ListItemText).prop('primary')).toBe('Login');
    });

    it('should call onShowLoginDialog when the button is clicked', () => {
      wrapper.find(ListItem).simulate('click');
      expect(mockShowLoginDialog).toHaveBeenCalledTimes(1);
    });
  });
});
