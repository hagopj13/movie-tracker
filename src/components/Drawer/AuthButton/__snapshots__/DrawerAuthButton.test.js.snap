// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DrawerAuthButton component isAuth is false should render the DrawerAuthButton component correctly 1`] = `
<WithStyles(ForwardRef(List))>
  <WithStyles(ForwardRef(ListItem))
    button={true}
    onClick={[MockFunction]}
  >
    <WithStyles(ForwardRef(ListItemIcon))>
      <LockOpenIcon />
    </WithStyles(ForwardRef(ListItemIcon))>
    <WithStyles(ForwardRef(ListItemText))
      primary="Login"
    />
  </WithStyles(ForwardRef(ListItem))>
</WithStyles(ForwardRef(List))>
`;

exports[`DrawerAuthButton component isAuth is true should render the DrawerAuthButton component correctly 1`] = `
<WithStyles(ForwardRef(List))>
  <WithStyles(ForwardRef(ListItem))
    button={true}
    onClick={[MockFunction]}
  >
    <WithStyles(ForwardRef(ListItemIcon))>
      <LockIcon />
    </WithStyles(ForwardRef(ListItemIcon))>
    <WithStyles(ForwardRef(ListItemText))
      primary="Logout"
    />
  </WithStyles(ForwardRef(ListItem))>
</WithStyles(ForwardRef(List))>
`;
